File: packages/ik_rules/lib/src/skills/level_unlocks.dart
import 'package:collection/collection.dart';
import 'package:ik_content/ik_content.dart';

import '../activity/requirements.dart';
import '../activity/xp.dart';
import '../js_compat.dart';
import '../production/recipes.dart';
import '../projects/projects.dart';
import '../recipes/knowledge.dart';
import '../save/generated/save_models.dart';

/// Activities, recipes, and projects that become available in a level range.
class SkillUnlockSummary {
  const SkillUnlockSummary({
    required this.unlockedActivities,
    required this.proficientActivities,
    required this.recipes,
    required this.projects,
  });

  final List<String> unlockedActivities;
  final List<String> proficientActivities;
  final List<String> recipes;
  final List<String> projects;

  bool get isEmpty =>
      unlockedActivities.isEmpty &&
      proficientActivities.isEmpty &&
      recipes.isEmpty &&
      projects.isEmpty;

  bool get isNotEmpty => !isEmpty;

  Map<String, Object?> toJson() => <String, Object?>{
    'unlockedActivities': unlockedActivities,
    'proficientActivities': proficientActivities,
    'recipes': recipes,
    'projects': projects,
  };
}

/// One skill that just rose, plus what that new level opened.
class SkillLevelUpNotice {
  const SkillLevelUpNotice({
    required this.skillId,
    required this.skillName,
    required this.level,
    required this.unlocks,
  });

  final String skillId;
  final String skillName;
  final num level;
  final SkillUnlockSummary unlocks;

  Map<String, Object?> toJson() => <String, Object?>{
    'skillId': skillId,
    'skillName': skillName,
    'level': level,
    'unlocks': unlocks.toJson(),
  };
}

bool _inLevelRange(num value, num fromLevel, num toLevel) {
  return value > fromLevel && value <= toLevel;
}

String _skillName(GameDatabase db, String skillId) {
  final name = db.skills.firstWhereOrNull((skill) => skill.skillId == skillId)?.displayName;
  return name is String && name.isNotEmpty ? name : skillId;
}

String _activityName(ActivityRow activity) {
  final contextual = activity.contextualName?.trim();
  if (contextual != null && contextual.isNotEmpty) return contextual;
  return activity.activityId;
}

List<ActionRow> _actionsForActivity(GameDatabase db, ActivityRow activity) {
  final poolId = activity.poolId;
  if (poolId == null || poolId.isEmpty) return const <ActionRow>[];
  final actionIds = <String>{
    for (final entry in db.poolEntries)
      if (entry.raw['Pool ID'] == poolId) jsString(entry.raw['Action ID']),
  };
  if (actionIds.isEmpty) return const <ActionRow>[];
  return [
    for (final action in db.actions)
      if (actionIds.contains(action.actionId)) action,
  ];
}

List<String> _uniqueSorted(Iterable<String> names) {
  final out = names.where((name) => name.trim().isNotEmpty).toSet().toList();
  mergeSort(out, compare: jsLocaleCompare);
  return out;
}

/// Names unlocked between [fromLevel] (exclusive) and [toLevel] (inclusive).
SkillUnlockSummary skillUnlocksBetween(
  GameDatabase db,
  PlayerSave saveAfter,
  String skillId,
  num fromLevel,
  num toLevel,
) {
  if (toLevel <= fromLevel) {
    return const SkillUnlockSummary(
      unlockedActivities: <String>[],
      proficientActivities: <String>[],
      recipes: <String>[],
      projects: <String>[],
    );
  }

  final unlocked = <String>[];
  final proficient = <String>[];

  for (final activity in db.activities) {
    if (activity.status == 'Needs Data') continue;
    if (!activityVisibleForSave(db, saveAfter, activity.activityId)) continue;
    final name = _activityName(activity);

    for (final requirement in requirementsForEntity(db, 'Activity', activity.activityId)) {
      if (requirement.requirementType != 'Skill Level') continue;
      if ((requirement.operatorValue ?? '').toLowerCase() == 'proficiency') continue;
      if (jsString(requirement.referenceIdValue) != skillId) continue;
      final required = requirement.requiredValue;
      if (required is num && _inLevelRange(required, fromLevel, toLevel)) {
        unlocked.add(name);
      }
    }

    for (final action in _actionsForActivity(db, activity)) {
      if (action.status == 'Needs Data') continue;
      if (action.relevantSkillId != skillId) continue;
      final proficiency = action.proficiencyLevel;
      if (proficiency is num && _inLevelRange(proficiency, fromLevel, toLevel)) {
        proficient.add(name);
      }
    }
  }

  final recipes = <String>[];
  for (final recipe in db.recipes.where(isCompleteRecipe)) {
    if (jsString(recipe.raw['Skill ID']) != skillId) continue;
    if (!knowsRecipe(saveAfter, db, jsString(recipe.raw['Recipe ID']))) continue;
    final proficiency = recipe.raw['Proficiency Level'];
    if (proficiency is num && _inLevelRange(proficiency, fromLevel, toLevel)) {
      recipes.add(jsString(recipe.raw['Display Name']));
    }
  }

  final projects = <String>[];
  for (final project in db.projects.where(isCompleteProject)) {
    if (!meetsProjectKnowledge(db, saveAfter, project)) continue;
    if (!meetsProjectSkills(saveAfter, project)) continue;
    final hit = projectSkillRequirements(project).any(
      (requirement) =>
          requirement.skillId == skillId && _inLevelRange(requirement.level, fromLevel, toLevel),
    );
    if (hit) projects.add(jsString(project.raw['Display Name']));
  }

  final unlockedNames = _uniqueSorted(unlocked);
  final unlockedSet = unlockedNames.toSet();
  final proficientNames = _uniqueSorted(proficient.where((name) => !unlockedSet.contains(name)));

  return SkillUnlockSummary(
    unlockedActivities: unlockedNames,
    proficientActivities: proficientNames,
    recipes: _uniqueSorted(recipes),
    projects: _uniqueSorted(projects),
  );
}

/// Skills that rose from [before] to [after], each with the unlocks in that jump.
List<SkillLevelUpNotice> skillLevelUpsBetween(
  GameDatabase db,
  PlayerSave before,
  PlayerSave after,
) {
  final fromLevels = <String, num>{for (final skill in before.skills) skill.skillId: skill.level};
  final notices = <SkillLevelUpNotice>[];
  final seen = <String>{};
  for (final skill in after.skills) {
    if (!seen.add(skill.skillId)) continue;
    final from = fromLevels[skill.skillId] ?? 1;
    if (skill.level <= from) continue;
    notices.add(
      SkillLevelUpNotice(
        skillId: skill.skillId,
        skillName: _skillName(db, skill.skillId),
        level: skill.level,
        unlocks: skillUnlocksBetween(db, after, skill.skillId, from, skill.level),
      ),
    );
  }
  mergeSort(notices, compare: (a, b) => jsLocaleCompare(a.skillName, b.skillName));
  return notices;
}

File: src/game/skills/levelUnlocks.ts
import { activityVisibleForSave, requirementsForEntity } from '../activity/requirements'
import type { ActionRow, ActivityRow, GameDatabase } from '../data/types'
import { isCompleteRecipe } from '../production/recipes'
import { isCompleteProject, meetsProjectKnowledge, meetsProjectSkills, projectSkillRequirements } from '../projects/projects'
import { knowsRecipe } from '../recipes/knowledge'
import type { PlayerSave } from '../save/types'

export interface SkillUnlockSummary {
  unlockedActivities: string[]
  proficientActivities: string[]
  recipes: string[]
  projects: string[]
}

export interface SkillLevelUpNotice {
  skillId: string
  skillName: string
  level: number
  unlocks: SkillUnlockSummary
}

function inLevelRange(value: number, fromLevel: number, toLevel: number): boolean {
  return value > fromLevel && value <= toLevel
}

function skillName(db: GameDatabase, skillId: string): string {
  const name = db.Skills.find((skill) => skill['Skill ID'] === skillId)?.['Display Name']
  return name ? name : skillId
}

function activityName(activity: ActivityRow): string {
  const contextual = typeof activity['Contextual Name'] === 'string' ? activity['Contextual Name'].trim() : ''
  return contextual || activity['Activity ID']
}

function actionsForActivity(db: GameDatabase, activity: ActivityRow): ActionRow[] {
  const poolId = activity['Pool ID']
  if (!poolId) return []
  const actionIds = new Set(
    db.PoolEntries.filter((entry) => entry['Pool ID'] === poolId).map((entry) => entry['Action ID']),
  )
  if (actionIds.size === 0) return []
  return db.Actions.filter((action) => actionIds.has(action['Action ID']))
}

function uniqueSorted(names: Iterable<string>): string[] {
  const out = [...new Set([...names].filter((name) => name.trim().length > 0))]
  out.sort((a, b) => a.localeCompare(b))
  return out
}

/** Names unlocked between `fromLevel` (exclusive) and `toLevel` (inclusive). */
export function skillUnlocksBetween(
  db: GameDatabase,
  saveAfter: PlayerSave,
  skillId: string,
  fromLevel: number,
  toLevel: number,
): SkillUnlockSummary {
  if (toLevel <= fromLevel) {
    return { unlockedActivities: [], proficientActivities: [], recipes: [], projects: [] }
  }

  const unlocked: string[] = []
  const proficient: string[] = []

  for (const activity of db.Activities) {
    if (activity.Status === 'Needs Data') continue
    if (!activityVisibleForSave(db, saveAfter, activity['Activity ID'])) continue
    const name = activityName(activity)

    for (const requirement of requirementsForEntity(db, 'Activity', activity['Activity ID'])) {
      if (requirement['Requirement Type'] !== 'Skill Level') continue
      if (String(requirement.Operator ?? '').toLowerCase() === 'proficiency') continue
      if (String(requirement['Reference ID / Value'] ?? '') !== skillId) continue
      const required = requirement['Required Value']
      if (typeof required === 'number' && inLevelRange(required, fromLevel, toLevel)) {
        unlocked.push(name)
      }
    }

    for (const action of actionsForActivity(db, activity)) {
      if (action.Status === 'Needs Data') continue
      if (action['Relevant Skill ID'] !== skillId) continue
      const proficiency = action['Proficiency Level']
      if (typeof proficiency === 'number' && inLevelRange(proficiency, fromLevel, toLevel)) {
        proficient.push(name)
      }
    }
  }

  const recipes: string[] = []
  for (const recipe of db.Recipes.filter(isCompleteRecipe)) {
    if (recipe['Skill ID'] !== skillId) continue
    if (!knowsRecipe(saveAfter, db, recipe['Recipe ID'])) continue
    const proficiency = recipe['Proficiency Level']
    if (typeof proficiency === 'number' && inLevelRange(proficiency, fromLevel, toLevel)) {
      recipes.push(recipe['Display Name'])
    }
  }

  const projects: string[] = []
  for (const project of db.Projects.filter(isCompleteProject)) {
    if (!meetsProjectKnowledge(db, saveAfter, project)) continue
    if (!meetsProjectSkills(saveAfter, project)) continue
    const hit = projectSkillRequirements(project).some(
      (requirement) =>
        requirement.skillId === skillId && inLevelRange(requirement.level, fromLevel, toLevel),
    )
    if (hit) projects.push(project['Display Name'])
  }

  const unlockedActivities = uniqueSorted(unlocked)
  const unlockedSet = new Set(unlockedActivities)
  return {
    unlockedActivities,
    proficientActivities: uniqueSorted(proficient.filter((name) => !unlockedSet.has(name))),
    recipes: uniqueSorted(recipes),
    projects: uniqueSorted(projects),
  }
}

/** Skills that rose from `before` to `after`, each with the unlocks in that jump. */
export function skillLevelUpsBetween(
  db: GameDatabase,
  before: PlayerSave,
  after: PlayerSave,
): SkillLevelUpNotice[] {
  const fromLevels = new Map(before.skills.map((skill) => [skill.skillId, skill.level] as const))
  const notices: SkillLevelUpNotice[] = []
  const seen = new Set<string>()
  for (const skill of after.skills) {
    if (seen.has(skill.skillId)) continue
    seen.add(skill.skillId)
    const from = fromLevels.get(skill.skillId) ?? 1
    if (skill.level <= from) continue
    notices.push({
      skillId: skill.skillId,
      skillName: skillName(db, skill.skillId),
      level: skill.level,
      unlocks: skillUnlocksBetween(db, after, skill.skillId, from, skill.level),
    })
  }
  notices.sort((a, b) => a.skillName.localeCompare(b.skillName))
  return notices
}
